pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -106,6 +120,53 @@ func haveOverlap(a1, a2 []string) bool {
 	return false
 }
 
+func (pl *VolumeRestrictions) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
+	return pl.isReadWriteOncePodAccessModeConflict(pod)
+}
+
+// isReadWriteOncePodAccessModeConflict checks if a pod uses a PVC with the ReadWriteOncePod access mode.
+// This access mode restricts volume access to a single pod on a single node. Since only a single pod can
+// use a ReadWriteOncePod PVC, mark any other pods attempting to use this PVC as unschedulable and unresolvable.
+func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(pod *v1.Pod) *framework.Status {
+	nodeInfos, err := pl.nodeInfoLister.NodeInfos().List()
+	if err != nil {
+		return framework.NewStatus(framework.Error, ErrReasonNodeInfoError)
+	}
+
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvc, err := pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(volume.PersistentVolumeClaim.ClaimName)
+		if err != nil {
+			return framework.NewStatus(framework.Error, ErrReasonPVCError)
+		}
+
+		// If ReadWriteOncePod is used, it can be the only one on the PVC since it overrides any
+		// additional access modes. This is also because we do not know what access mode the pod
+		// intends to use if there are multiple.
+		if !(v1helper.ContainsAccessMode(pvc.Spec.AccessModes, v1.ReadWriteOncePod) && len(pvc.Spec.AccessModes) == 1) {
+			continue
+		}
+
+		key := pod.Namespace + "/" + volume.PersistentVolumeClaim.ClaimName
+		for _, nodeInfo := range nodeInfos {
```

Comment:
we should run this in parallel

Reply:
this is still valid, we need to run it in parallel, see the below example from the PreFilter in pod affinity: https://github.com/kubernetes/kubernetes/blob/41c3f5b58db08b2a1d17a2ebecd6cc6d3cb89086/pkg/scheduler/framework/plugins/interpodaffinity/filtering.go#L175

Reply:
Thank you for the clarification, done